File: src/audit_reports/document_origin.py
"""Compare a fresh official download with acquired bytes; never replace either.

Matching bytes establish source revision agreement, not text/table accuracy.
Different revisions and unavailable URLs remain explicit review outcomes.
"""
from __future__ import annotations

from datetime import datetime, timezone
import hashlib
import json
from pathlib import Path
import re
from urllib.parse import urlsplit

import fitz

from .document_acquisition import fetch_source, unwrap_pdf
from .document_corpus import Filing
from .document_corpus_resume import metadata
from .document_corpus_store import PREFIX, _error_code, _json
from .document_quality import source_identity_review


def _sha(body):
    return hashlib.sha256(body).hexdigest()


def _validate_bound_selection(entry):
    if not isinstance(entry, dict):
        raise ValueError('Invalid source-bound archive selection')
    filing = Filing(**entry['filing'])
    url = urlsplit(entry['source_url'])
    if url.scheme != 'https' or not url.hostname or url.username or url.password or url.fragment:
        raise ValueError('Archive selection requires its exact HTTPS source URL')
    if (entry.get('schema_version') != 'document-origin-selection-1'
            or entry.get('semantically_verified') is not False
            or not isinstance(entry.get('member'), str) or not entry['member'].lower().endswith('.pdf')):
        raise ValueError('Invalid source-bound archive selection')
    review = entry['review']
    digests = [entry['transport_sha256'], entry['sha256'], review['origin_observation_sha256']]
    if review['basis'] == 'acquired_pdf_byte_agreement':
        digests.append(review['acquisition_sha256'])
        if review['acquisition_sha256'] != entry['sha256']:
            raise ValueError('Selected archive member differs from the reviewed acquisition')
    elif review['basis'] == 'rendered_financial_report_cover':
        if (review.get('page') != 1 or not isinstance(review.get('text'), str)
                or not review['text'].strip()):
            raise ValueError('Reviewed report selection requires its literal cover text')
    else:
        raise ValueError('Unknown archive selection review basis')
    if any(not isinstance(value, str) or not re.fullmatch(r'[a-f0-9]{64}', value) for value in digests):
        raise ValueError('Invalid archive selection source digest')
    return filing, entry['source_url']


def load_archive_selections(path: Path, registered) -> dict:
    """A separate official URL needs its own reviewed, exact-byte selection."""
    packet = json.loads(path.read_text(encoding='utf-8'))
    if packet.get('schema_version') != 'document-origin-selections-1' or not isinstance(packet.get('selections'), list):
        raise ValueError('Invalid official archive selection registry')
    result = {}
    for entry in packet['selections']:
        binding = _validate_bound_selection(entry)
        if binding[0] not in registered or binding in result:
            raise ValueError('Unregistered or duplicate official archive selection')
        result[binding] = entry
    return result


def observe_origin(store, filing: Filing, acquisition_key: str | None, url: str, patterns: dict, *,
                   reviewed_member: dict | None = None, fetch=fetch_source, checked_at: str | None = None,
                   listing: tuple[bytes, dict] | None = None):
    result = {'schema_version': 'document-origin-review-1', 'filing': filing.as_dict(), 'source_url': url,
              'acquisition_key': acquisition_key, 'requested_archive_selection': reviewed_member,
              'checked_at': checked_at or datetime.now(timezone.utc).isoformat(),
              'acquisition': None, 'transport': None, 'origin_pdf': None,
              'semantically_verified': False,
              'engine': {'pymupdf': fitz.VersionBind, 'implementation': {
                  name: _sha(Path(__file__).with_name(name).read_bytes()) for name in
                  ('document_origin.py', 'document_acquisition.py', 'document_quality.py')}}}
    acquired = None
    if acquisition_key:
        try:
            response = store.client.get_object(Bucket=store.bucket, Key=acquisition_key)
        except Exception as error:
            if _error_code(error) not in ('404', 'NoSuchKey'):
                raise
        else:
            try:
                acquired = response['Body'].read()
            finally:
                response['Body'].close()
            if len(acquired) != response['ContentLength']:
                raise ValueError('Acquisition read was truncated during origin comparison')
            result['acquisition'] = {'key': acquisition_key, 'sha256': _sha(acquired), 'bytes': len(acquired),
                                     'version': metadata(acquisition_key, response)}
    artifacts = {}
    if listing is not None:
        from .document_origin_listing import listing_witness
        witness = listing_witness(*listing, filing)
        if witness['download_url'] != url:
            raise ValueError('Origin URL differs from its official listing row')
        result['source_listing'] = witness
        artifacts['source_listing'] = listing[0]
    try:
        transport, response = fetch(url)
    except Exception as error:
        result.update(status='origin_unavailable', error=str(error))
        return result, artifacts
    result.update(response=response, transport={'sha256': _sha(transport), 'bytes': len(transport)})
    artifacts['transport'] = transport
    try:
        bound_selection = reviewed_member and any(k in reviewed_member for k in ('source_url', 'transport_sha256'))
        if bound_selection:
            if _validate_bound_selection(reviewed_member) != (filing, url):
                raise ValueError('Reviewed archive selection belongs to a different filing or source URL')
            review = reviewed_member['review']
            if (review['basis'] == 'acquired_pdf_byte_agreement'
                    and (acquired is None or _sha(acquired) != review['acquisition_sha256'])):
                raise ValueError('Acquisition differs from the reviewed archive selection')
        body, selection = unwrap_pdf(transport, reviewed_member)
        result['selection'] = selection
        result['origin_pdf'] = {'sha256': _sha(body), 'bytes': len(body)}
        artifacts['origin_pdf'] = body
        leading = []
        with fitz.open(stream=body, filetype='pdf') as pdf:
            if pdf.needs_pass or not len(pdf):
                raise ValueError('Origin PDF requires a password or has no pages')
            for number in range(min(3, len(pdf))):
                raw = pdf[number].get_text('dict', flags=fitz.TEXTFLAGS_DICT & ~fitz.TEXT_PRESERVE_IMAGES,
                                           clip=fitz.INFINITE_RECT())
                spans = []
                for block_id, block in enumerate(raw['blocks']):
                    if block['type'] != 0:
                        continue
                    for line_id, line in enumerate(block['lines']):
                        for span in line['spans']:
                            spans.append({'id': len(spans), 'block': block_id, 'line': line_id,
                                          'text': span['text'], 'bbox': list(span['bbox'])})
                leading.append({'page': number + 1, 'spans': spans})
            result['page_count'] = len(pdf)
        result['origin_leading_pages'] = leading
        if bound_selection and reviewed_member['review']['basis'] == 'rendered_financial_report_cover':
            cover = ' '.join(s['text'] for s in leading[0]['spans'])
            if ' '.join(cover.split()) != ' '.join(reviewed_member['review']['text'].split()):
                raise ValueError('Selected PDF cover differs from its source review')
        result['origin_identity'] = source_identity_review(filing, leading, patterns)
        result['related_pdf_content_capture'] = 'pending' if selection.get('unselected_pdf_members') else 'not_applicable'
    except Exception as error:
        result.update(status='origin_needs_review', error=str(error))
        return result, artifacts
    if acquired is None:
        status = 'acquisition_missing'
    elif acquired == body:
        status = 'matches_acquired_bytes'
    else:
        status = 'different_pdf_revision'
        # Only an observed serialized wrapper may be normalized. Metadata,
        # punctuation, PDF objects and visible contents are never discarded.
        if acquired.startswith(b'\xac\xed\x00\x05'):
            try:
                normalized, wrapper = unwrap_pdf(acquired)
            except ValueError as error:
                result['acquisition_wrapper_error'] = str(error)
            else:
                result['acquisition_wrapper'] = wrapper
                if normalized == body:
                    status = 'same_pdf_after_acquisition_wrapper'
    result['status'] = status
    return result, artifacts


def publish_origin(store, result: dict, artifacts: dict[str, bytes], patterns: dict) -> dict:
    """Keep downloaded evidence and an immutable review before updating its index."""
    filing = Filing(**result['filing'])
    if result.get('schema_version') != 'document-origin-review-1' or result.get('semantically_verified') is not False:
        raise ValueError('Invalid source-origin review')
    observed_at = datetime.fromisoformat(result['checked_at'])
    if observed_at.tzinfo is None or observed_at.utcoffset().total_seconds() != 0:
        raise ValueError('Origin observation time must have an explicit UTC offset')
    acquired = result['acquisition']
    if acquired:
        current = metadata(acquired['key'], store.client.head_object(Bucket=store.bucket, Key=acquired['key']))
        if current != acquired['version']:
            raise ValueError('Acquisition changed during origin review')
    expected = {name for name in ('transport', 'origin_pdf', 'source_listing') if result.get(name) is not None}
    if set(artifacts) != expected:
        raise ValueError('Origin review is missing its downloaded evidence')
    def retained_download(_url):
        if 'transport' not in artifacts:
            raise RuntimeError(result['error'])
        return artifacts['transport'], result['response']
    checked, retained = observe_origin(store, filing, result['acquisition_key'], result['source_url'], patterns,
        reviewed_member=result['requested_archive_selection'], fetch=retained_download, checked_at=result['checked_at'],
        listing=(artifacts['source_listing'], result['source_listing']['response']) if 'source_listing' in artifacts else None)
    if checked != result or retained != artifacts:
        raise ValueError('Origin review differs from acquired bytes and retained transport')
    value = dict(result)
    for name, body in artifacts.items():
        record = result[name]
        if _sha(body) != record['sha256'] or len(body) != record['bytes']:
            raise ValueError('Origin artifact differs from its observed bytes')
        if name == 'source_listing':
            key = f"{PREFIX}origin-listings/{record['sha256']}.html"
        else:
            key = (f"{PREFIX}transports/{record['sha256']}/original.bin" if name == 'transport' else
                   f"{PREFIX}sources/{record['sha256']}/original.pdf")
        store._immutable(key, body, 'application/pdf' if name == 'origin_pdf' else 'application/octet-stream')
        value[name] = {**record, 'key': key}
    base = f'{PREFIX}origins/{filing.bank_ticker}/{filing.period}/{filing.kind}/'
    payload = _json(value)
    digest = _sha(payload)
    key = base + digest + '.json'
    store._immutable(key, payload, 'application/json')
    revision = {'key': key, 'sha256': digest, 'bytes': len(payload), 'checked_at': value['checked_at'],
                'status': value['status'], 'acquisition_sha256': acquired['sha256'] if acquired else None}
    for _ in range(8):
        previous, etag = store._read(base + 'index.json')
        index = json.loads(previous) if previous else {'schema_version': 'document-origin-index-1',
            'filing': filing.as_dict(), 'current': None, 'revisions': [], 'semantically_verified': False}
        if (index['filing'] != filing.as_dict() or index['schema_version'] != 'document-origin-index-1'
                or index.get('semantically_verified') is not False):
            raise ValueError('Origin index differs from its filing')
        if revision not in index['revisions']:
            index['revisions'].append(revision)
        index['current'] = max(index['revisions'], key=lambda r: (r['checked_at'], r['sha256']))
        body = _json(index)
        if body == previous:
            return {**value, 'review_key': key, 'index_key': base + 'index.json'}
        try:
            store.client.put_object(Bucket=store.bucket, Key=base + 'index.json', Body=body,
                                    ContentType='application/json', **({'IfMatch': etag} if etag else {'IfNoneMatch': '*'}))
            readback, _ = store._read(base + 'index.json')
            retained_index = json.loads(readback)
            if (retained_index.get('filing') != filing.as_dict()
                    or retained_index.get('schema_version') != 'document-origin-index-1'
                    or retained_index.get('semantically_verified') is not False
                    or revision not in retained_index.get('revisions', [])):
                raise ValueError('Origin index readback does not retain this observation')
            return {**value, 'review_key': key, 'index_key': base + 'index.json'}
        except Exception as error:
            if _error_code(error) not in ('412', 'PreconditionFailed'):
                raise
    raise RuntimeError('Origin index changed concurrently; retry this review')

File: web/app/lib/document-origin.ts
/** Independent official-source observations; byte agreement never approves contents. */
import { CORPUS_PREFIX, type CorpusBucket, type FilingIdentity } from "./document-corpus";
import { readRecoveryArtifact } from "./document-recovery";
import regulatorNames from "../../../data/banks/bddk_audit_registry_names.json";

export type OriginReference = { key: string; sha256: string; bytes: number; checked_at: string; status: string; acquisition_sha256: string | null };
type Artifact = { key: string; sha256: string; bytes: number };
export type OriginObservation = { reference: OriginReference; current: boolean; review: OriginReview };
export type OriginReview = {
  schema_version: "document-origin-review-1"; filing: FilingIdentity; checked_at: string;
  status: "matches_acquired_bytes" | "same_pdf_after_acquisition_wrapper" | "different_pdf_revision"
    | "acquisition_missing" | "origin_unavailable" | "origin_needs_review";
  source_url: string; semantically_verified: false; error?: string;
  acquisition: { sha256: string; bytes: number } | null;
  transport: Artifact | null; origin_pdf: Artifact | null;
  source_listing?: Artifact & { registry_name: string; download_url: string; row_number: number;
    row_cells: string[]; response: { source_url: string } };
  origin_identity?: { status: string }; related_pdf_content_capture?: string;
  selection?: { archive_member?: string; unselected_pdf_members?: { name: string; bytes: number; sha256: string }[] };
};
const record = (v: unknown): v is Record<string, unknown> => typeof v === "object" && v !== null && !Array.isArray(v);
const hash = (v: unknown): v is string => typeof v === "string" && /^[a-f0-9]{64}$/.test(v);
const count = (v: unknown): v is number => typeof v === "number" && Number.isSafeInteger(v) && v > 0;
const sameFiling = (v: unknown, f: FilingIdentity) => record(v) && Object.entries(f).every(([k, value]) => v[k] === value);
const time = (v: unknown): v is string => typeof v === "string" && /(?:Z|\+00:00)$/.test(v) && Number.isFinite(Date.parse(v));
const statuses = new Set(["matches_acquired_bytes", "same_pdf_after_acquisition_wrapper", "different_pdf_revision",
  "acquisition_missing", "origin_unavailable", "origin_needs_review"]);

async function getOriginIndex(bucket: CorpusBucket, filing: FilingIdentity) {
  const base = `${CORPUS_PREFIX}origins/${filing.bank_ticker}/${filing.period}/${filing.kind}/`;
  const object = await bucket.get(base + "index.json");
  if (!object) return null;
  if (object.size > 4_000_000) throw new Error("Oversized origin index");
  const index: unknown = await object.json();
  if (!record(index) || index.schema_version !== "document-origin-index-1" || !sameFiling(index.filing, filing)
      || index.semantically_verified !== false || !Array.isArray(index.revisions) || !record(index.current)) {
    throw new Error("Invalid origin index binding");
  }
  const fields = ["key", "sha256", "bytes", "checked_at", "status", "acquisition_sha256"];
  if (!index.revisions.length || index.revisions.length > 1000
      || new Set(index.revisions.map(r => record(r) ? r.sha256 : null)).size !== index.revisions.length
      || !index.revisions.every(r => record(r) && hash(r.sha256) && r.key === `${base}${r.sha256}.json`
        && count(r.bytes) && r.bytes <= 8_000_000 && time(r.checked_at) && statuses.has(String(r.status))
        && (r.acquisition_sha256 === null || hash(r.acquisition_sha256)))
      || !index.revisions.some(r => record(r) && fields.every(k => r[k] === (index.current as Record<string, unknown>)[k]))) {
    throw new Error("Invalid origin revision history");
  }
  return { current: index.current as OriginReference, revisions: index.revisions as OriginReference[] };
}

export async function getOriginReview(bucket: CorpusBucket, filing: FilingIdentity, observation?: string): Promise<OriginReview | null> {
  if (observation !== undefined && !hash(observation)) throw new Error("Invalid origin observation digest");
  const index = await getOriginIndex(bucket, filing);
  if (!index) return null;
  const reference = observation ? index.revisions.find(r => r.sha256 === observation) : index.current;
  if (!reference) throw new Error("Observation is absent from retained origin history");
  return readOriginReview(bucket, filing, reference);
}

export async function getOriginObservations(bucket: CorpusBucket, filing: FilingIdentity): Promise<OriginObservation[]> {
  const index = await getOriginIndex(bucket, filing);
  if (!index) return [];
  const results: OriginObservation[] = [];
  for (const reference of [...index.revisions].sort((a, b) => b.checked_at.localeCompare(a.checked_at) || b.sha256.localeCompare(a.sha256))) {
    results.push({ reference, current: reference.sha256 === index.current.sha256, review: await readOriginReview(bucket, filing, reference) });
  }
  return results;
}

async function readOriginReview(bucket: CorpusBucket, filing: FilingIdentity, current: OriginReference): Promise<OriginReview> {
  const bytes = await readRecoveryArtifact(bucket, current as Artifact);
  const value: unknown = JSON.parse(new TextDecoder("utf-8", { fatal: true }).decode(bytes));
  if (!record(value) || value.schema_version !== "document-origin-review-1" || !sameFiling(value.filing, filing)
      || value.semantically_verified !== false || value.status !== current.status || value.checked_at !== current.checked_at
      || typeof value.source_url !== "string" || !/^https?:\/\//.test(value.source_url)
      || value.error !== undefined && typeof value.error !== "string") throw new Error("Origin receipt binding mismatch");
  if (value.acquisition !== null && (!record(value.acquisition) || !hash(value.acquisition.sha256)
      || !count(value.acquisition.bytes))) throw new Error("Invalid acquired-source observation");
  if ((record(value.acquisition) ? value.acquisition.sha256 : null) !== current.acquisition_sha256) {
    throw new Error("Origin acquisition differs from its index");
  }
  for (const name of ["transport", "origin_pdf"]) {
    const entry = value[name];
    if (entry === null) continue;
    if (!record(entry) || !hash(entry.sha256) || !count(entry.bytes)
        || entry.key !== (name === "transport" ? `${CORPUS_PREFIX}transports/${entry.sha256}/original.bin`
          : `${CORPUS_PREFIX}sources/${entry.sha256}/original.pdf`)) throw new Error("Invalid origin artifact key");
  }
  if (value.source_listing !== undefined) {
    const listing = value.source_listing;
    const expectedName = (regulatorNames.banks as Record<string, string>)[filing.bank_ticker];
    if (!record(listing) || listing.schema_version !== "document-origin-listing-1" || !sameFiling(listing.filing, filing)
        || listing.semantically_verified !== false || !hash(listing.sha256) || !count(listing.bytes)
        || listing.key !== `${CORPUS_PREFIX}origin-listings/${listing.sha256}.html` || listing.bytes > 20_000_000
        || !expectedName || listing.registry_name !== expectedName || listing.download_url !== value.source_url
        || !record(listing.engine) || !hash(listing.engine.implementation_sha256) || !hash(listing.engine.bank_names_sha256)
        || typeof listing.row_number !== "number" || !Number.isSafeInteger(listing.row_number) || listing.row_number < 0
        || !Array.isArray(listing.row_cells) || listing.row_cells.length !== 5
        || listing.row_cells[0] !== expectedName || listing.row_cells[1] !== filing.period.slice(0, 4)
        || listing.row_cells[2] !== String(Number(filing.period.slice(-1)) * 3)
        || listing.row_cells[3] !== (filing.kind === "consolidated" ? "KONSOLIDE" : "SOLO")
        || !record(listing.response) || listing.response.source_url !== "https://www.bddk.org.tr/BdrUyg/"
        || listing.response.resolved_url !== "https://www.bddk.org.tr/BdrUyg/Home/SorguSonuc?KurulusTuru=1&EFTKodu=0&RaporTipi=T%C3%9CM%C3%9C&DonemYil=0&DonemAy=0"
        || listing.response.method !== "POST"
        || !record(listing.response.form) || listing.response.form.KurulusTuru !== "1" || listing.response.form.EFTKodu !== "0"
        || listing.response.form.RaporTipi !== "TÜMÜ" || listing.response.form.DonemYil !== "0" || listing.response.form.DonemAy !== "0") {
      throw new Error("Invalid regulator listing witness");
    }
    const url = new URL(value.source_url as string);
    const target = url.searchParams.getAll("raporUrl");
    const basis = filing.kind === "consolidated" ? "KONSOLIDE" : "SOLO";
    const match = /^~\/Dosya\/BDREki-\d+-(SOLO|KONSOLIDE)-(\d{4})-(\d{2})\.zip$/.exec(target[0] ?? "");
    if (url.protocol !== "https:" || url.host !== "www.bddk.org.tr" || url.pathname !== "/BdrUyg/Home/DosyaIndir"
        || url.hash || [...url.searchParams.keys()].length !== 1 || target.length !== 1
        || !match || match[1] !== basis || match[2] !== filing.period.slice(0, 4)
        || match[3] !== String(Number(filing.period.slice(-1)) * 3).padStart(2, "0")) throw new Error("Regulator link differs from its filing");
  }
  if (value.origin_identity !== undefined && (!record(value.origin_identity) || typeof value.origin_identity.status !== "string")) {
    throw new Error("Invalid origin identity observation");
  }
  if (value.selection !== undefined) {
    if (!record(value.selection)) throw new Error("Invalid origin archive selection");
    const members = value.selection.unselected_pdf_members;
    if (members !== undefined && (!Array.isArray(members) || members.some(m => !record(m)
        || typeof m.name !== "string" || !count(m.bytes) || !hash(m.sha256)))) throw new Error("Invalid related PDF members");
  }
  const acquiredHash = record(value.acquisition) ? value.acquisition.sha256 : null;
  const originHash = record(value.origin_pdf) ? value.origin_pdf.sha256 : null;
  if (value.status === "matches_acquired_bytes" && (!acquiredHash || acquiredHash !== originHash)
      || value.status === "different_pdf_revision" && (!acquiredHash || !originHash || acquiredHash === originHash)
      || value.status === "acquisition_missing" && (acquiredHash !== null || !originHash)
      || value.status === "origin_unavailable" && (value.transport !== null || value.origin_pdf !== null)
      || value.status === "same_pdf_after_acquisition_wrapper" && (!acquiredHash || !originHash || !record(value.acquisition_wrapper))) {
    throw new Error("Origin comparison contradicts observed source hashes");
  }
  return value as unknown as OriginReview;
}
